worker/peergrouper/desired.go

```diff
@@ -173,40 +192,45 @@ func desiredPeerGroup(info *peerGroupInfo) (bool, map[string]*replicaset.Member,
 	//    are not eligible to be primary.
 	// 3) Remove them.
 	// 4) Do nothing.
-	err := peerChanges.checkExtraMembers(info.extra)
+	err := p.checkExtraMembers()
 	if err != nil {
-		return false, nil, nil, errors.Trace(err)
+		return desiredChanges{}, errors.Trace(err)
 	}
 
-	peerChanges.members = info.initNewReplicaSet()
-	peerChanges.possiblePeerGroupChanges(info)
-	peerChanges.reviewPeerGroupChanges(info)
-	peerChanges.createNonVotingMember(&info.maxMemberId)
+	p.desired.members = p.info.initNewReplicaSet()
```

Comment:
Since the big effect is to clean up the high level API, I think I would take initNewReplicaSet off peerGroupInfo and tuck it (with the assignment) behind peerGroupChanges.